### PR TITLE
Arm: Add soc support for rda5981a 

### PR DESCRIPTION
**Pull request by liliu-l**

### CODE REVIEW

**Comment by nashif** [comment]
Can you add some more details in the commit message what rda5991h is exactly, a link maybe, which vendor is this, etc... Do not expect everyone to know what this board is.

**Comment by liliu-l** [comment]
commit message updated.

**Review by JammyZhou — COMMENTED** [review]
How about change the folder name "96b_rda" to "96b_bluesky" since the board is announced as BlueSky at the Linaro BUD17 connect? Accordingly, rename the '96b_rda_defconfig' file to '96b_bluesky_defconfig' as well as other macro names, etc

File: boards/arm/96b_rda/96b_rda_defconfig [hunk 1]
```diff
@@ -0,0 +1,27 @@
+CONFIG_ARM=y
+CONFIG_BOARD_96B_RDA=y
+CONFIG_SOC_SERIES_RDA5991=y
```
Comment (JammyZhou) [inline boards/arm/96b_rda/96b_rda_defconfig]: Should it be changed to CONFIG_SOC_SERIES_RDA5981?

**Comment by liliu-l** [comment]
changed 96b_rda to 96b_bluesky

**Comment by liliu-l** [comment]
resolved conflicts

File: arch/arm/soc/rda_5981/Kconfig [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#
+# Copyright (c) 2017 RDA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config SOC_FAMILY_RDA5981
```
Comment (MaureenHelm) [inline arch/arm/soc/rda_5981/Kconfig]: This is a really specific name for an SoC family. Is there a more general name you could use that could encompass additional SoCs in the family?

File: arch/arm/soc/rda_5981/Kconfig [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+if SOC_FAMILY_RDA5981
+config SOC_FAMILY
+	string
+	default rda_5981
+
+config SOC_RDA_DEBUG
+	bool
+	prompt "Use for RDA SOC DBG"
+	default n
```
Comment (MaureenHelm) [inline arch/arm/soc/rda_5981/Kconfig]: Can you please explain this more? Is this hardware debug?

File: arch/arm/soc/rda_5981/Kconfig.defconfig [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+source "arch/arm/soc/rda_5981/*/Kconfig.defconfig.series"
+
+if SOC_FAMILY_RDA5981
+
+config SRAM_BASE_ADDRESS
+	default 0x100080
+
+config FLASH_BASE_ADDRESS
+	default 0x018001000
```
Comment (MaureenHelm) [inline arch/arm/soc/rda_5981/Kconfig.defconfig]: sram and flash base addresses should come from DTS

File: arch/arm/soc/rda_5981/5981a/soc_registers.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+#ifndef _RDA5981a_SOC_REGISTERS_H_
+#define _RDA5981a_SOC_REGISTERS_H_
+
+/*------------- Watchdog Timer (WDT) ---------------------------------------*/
+struct wdt_rda5981a {
+	uint32_t wdt_cfg;
+};
+
+/* include register mapping headers */
+struct uart_rda5981a {
```
Comment (MaureenHelm) [inline arch/arm/soc/rda_5981/5981a/soc_registers.h]: Does this SoC have CMSIS headers that we could use instead?

File: arch/arm/soc/rda_5981/5981a/soc_cfg.c [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+	if (0 == (ch >> 1)) {
+		rd_rf_usb_reg(0xb2, &val, 0);
+		val &= ~(0x3 << 8);
+		wr_rf_usb_reg(0xb2, (val | (0x1 << (9 - ch))), 0);
+	}
+}
+
+/* Read ADC value */
+unsigned short rda_ccfg_adc_read(unsigned char ch)
```
Comment (MaureenHelm) [inline arch/arm/soc/rda_5981/5981a/soc_cfg.c]: This should go in a separate ADC driver

File: arch/arm/soc/rda_5981/5981a/soc_cfg.c [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+	while (RF_SPI_REG & (0x01 << 31))
+		;
+	while (RF_SPI_REG & (0x01 << 31))
+		;
+	*d = (unsigned short)(RF_SPI_REG & 0xFFFF);
+}
+
+/* Config UART0 RX pin */
+static inline int rda_ccfg_gp26(void)
```
Comment (MaureenHelm) [inline arch/arm/soc/rda_5981/5981a/soc_cfg.c]: Do you have a UART driver?

File: drivers/gpio/gpio_rda5981a.c [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+	gpio_rda5981a_config(dev, GPIO_ACCESS_BY_PIN, GPIO_PIN3,
+				(GPIO_DIR_OUT));
+
+	while (1) {
+		gpio_rda5981a_write(dev, GPIO_ACCESS_BY_PIN,
+					GPIO_PIN2, toggle);
+		gpio_rda5981a_write(dev, GPIO_ACCESS_BY_PIN,
+					GPIO_PIN3, !toggle);
+		k_sleep(1000);
+		toggle = toggle ? 0 : 1;
```
Comment (MaureenHelm) [inline drivers/gpio/gpio_rda5981a.c]: Why is this splash thread in the gpio driver? It looks more like a sample or test application

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Need to add board docs and DTS.

It also looks like you need to rebase to the latest arm branch.

File: arch/arm/soc/rda_5981/5981a/soc_cfg.c (hunk 7 above)
Comment (liliu-l) [inline arch/arm/soc/rda_5981/5981a/soc_cfg.c]: yes, we have UART driver, will push UART driver later.

File: arch/arm/soc/rda_5981/Kconfig (hunk 2 above)
Comment (liliu-l) [inline arch/arm/soc/rda_5981/Kconfig]: There is not yet a more generic name. 5981 is the SoC  famliy name, which includes 5981a now.

File: arch/arm/soc/rda_5981/Kconfig.defconfig (hunk 4 above)
Comment (liliu-l) [inline arch/arm/soc/rda_5981/Kconfig.defconfig]: DTS is not  defined currently, will add DTS later.

File: arch/arm/soc/rda_5981/5981a/soc_cfg.c (hunk 6 above)
Comment (liliu-l) [inline arch/arm/soc/rda_5981/5981a/soc_cfg.c]: will move this to ADC driver.

File: arch/arm/soc/rda_5981/Kconfig (hunk 3 above)
Comment (liliu-l) [inline arch/arm/soc/rda_5981/Kconfig]: It is not hardware debug.If select Y here, you can  print more debugging messages.

File: drivers/gpio/gpio_rda5981a.c (hunk 8 above)
Comment (liliu-l) [inline drivers/gpio/gpio_rda5981a.c]: It is s test case.  Have removed it.

Comment (liliu-l) [inline drivers/gpio/gpio_rda5981a.c]: It is s test case, have removed it.

**Comment by galak** [comment]
Can you rebase this against recent arm branch, and address the issues that shippable reported

File: arch/arm/soc/rda_5981/5981a/soc_registers.h (hunk 5 above)
Comment (liliu-l) [inline arch/arm/soc/rda_5981/5981a/soc_registers.h]: Yes, rda5981 has CMSIS headers, will add it later.

**Comment by galak** [comment]
Closing out as this seems abandoned.

**Pull request closed without merging**